lib/Sema/CSDiag.cpp

```diff
@@ -5421,8 +5562,11 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
     // Type check the function subexpression to resolve a type for it if
     // possible.
     fnExpr = typeCheckChildIndependently(callExpr->getFn());
-    if (!fnExpr)
-      return true;
+    // If we are misusing a subscript, provide a fixit.
+    if (!fnExpr) {
+      diagnoseSubscriptMisuse(callExpr, CS, *this);
```

Comment:
it would be great to change this to `if (diagnoseSubscriptMisuse(callExpr, CS, *this)) return true;` so it this case we don't have to query diagnostic engine.

Reply:
Good remark on the previous ones. Here though, if we change it this way, we will change the original flow. Originally, it was 
```
if (!fn)
  // the fixit was here
  return true;
```
You are proposing 
```
if (!fn) 
  if (diagnoseSubscriptMisuse(callExpr, CS, *this)) return true;
  // diagnoseSubscriptMisuse will most likely return false, but we want to 
  // always return true or hadAnyError()
```

Reply:
How about this, maybe `diagnoseSubscriptMisuse` should be decoupled from checking fo the functor:

```c++
if (diagnoseSubscriptMisuse(callExpr, CS, *this))
  return true;

fnExpr = typeCheckChildIndependently(callExpr->getFn());
if (!fnExpr)
  return CS.TC.Diags.hadAnyError();
```

Because it seems like diagnostic should be dependent on the subsequent check.

I was just thinking if we leave it this way there is no point of returning anything from `diagnoseSubscriptMisuse` because that gets ignored anyway... Maybe it should be switched back to return `void` this case.

Reply:
The thing is `diagnoseSubscriptMisuse` is dependent on the `if (!fn)` condition because type-checking has to fail and additionally I am expecting  and checking for the `has no member` error. Remember the talk about context lacking? The error that I added happens in `checkUnviableLookupResults`, but I could diagnose the note and fixit together with it for obvious reasons. I'm grabbing it here under `if (!fn)` for that very reason

Reply:
> because type-checking has to fail and additionally I am expecting and checking for the has no member error.

Remind me why that is? It seems like condition inside `diagnoseSubscriptMisuse` should be valid independently from `typeCheck` on functor, it might produce a different diagnostic but that's not really important. It would actually be great if we only produced "subscript" error and not "no such member" because "subscript" error is no actionable.

Reply:
Firstly, we can't produce this note if there is no error. The error comes from `typeCheckChildIndependently(callExpr->getFn())`, IIRC the call chain is something like `diagnoseMemberFailures -> ... -> diagnoseUnviableLookupResults`. When the error comes, we check `fnExpr` against `nullptr` and return true if it is. Before returning true, I check if we are dealing with a `subscript` member error and add the note with the fixit.
> It would actually be great if we only produced "subscript" error and not "no such member" because "subscript" error is no actionable.
What precisely is the 'subscript' error?

Reply:
> What precisely is the 'subscript' error?

The error you've added - `could_not_find_subscript_member`, I think you can just produce it right in `diagnoseSubscriptMisuse` and attach notes to it, because notes could only be attached if we know exactly what happened right? It doesn't seem to be a reasonable to depend on typecheck + `diagnoseUnviableLookupResults` to diagnose it, but it's good that `diagnoseUnviableLookupResults` can produce it independently or `diagnoseSubscriptMisuse` should be moved to that method altogether.

Reply:
To produce it not depending on `diagnoseUnviableLookupResults` I would have to do the work the latter method does actually diagnose the error. Right now I'm only diagnosing the note in `diagnoseSubscriptMisuse`.. 

> or diagnoseSubscriptMisuse should be moved to that method altogether.

I didn't quite understand the last part, move `diagnoseSubscriptMisuse` to `diagnoseUnviableLookupResults`?

Reply:
Yes, that's right, maybe it should just be moved to `diagnoseUnviableLookupResults`  instead of `visitApplyExpr` but probably it can't happen because `baseExpr` there most likely going to be `UnresolvedDotExpr` hence no access to the arguments :(

Reply:
Yes, that is exactly why I am diagnosing in `visitApplyExpr`.

Reply:
Makes sense, try emitting error in `diagnoseSubscriptMisuse` and see what happens.

Reply:
What I mean is trying switching to:

```c++
if (diagnoseSubscriptMisuse(callExpr, CS, *this))
  return true;

fnExpr = typeCheckChildIndependently(callExpr->getFn());
if (!fnExpr)
  return CS.TC.Diags.hadAnyError();
```

and emit error directly in `diagnoseSubscriptMisuse` and see what is going to break. if fix-its can't be produced it would just fallthrough into `typeCheckChildIndependently` and get diagnosed anyway, otherwise one call to type-check is going to be saved.

Reply:
If I understand correctly, you want the error to be emitted together with the note in `diagnoseSubscriptMisuse`, but only if the note is diagnosed and the fixits are valid. 
You stated the advantage. The downsides are that we diagnose one thing in two different places and in the case when the fixits can't be produced, we not only fallthrough into `typeCheckChildIndependently`, but have also already done the member lookup, `typeCheckChildIndependently(argExpr)` and argument compatibility checks. In other words, we do twice as much work than we could have done, if not more, to just diagnose the error without the fixits.
Honestly I think the best thing we can do is void the function and leave everything as is. It is quite safe because all errors related to arguments are out the scope for `typeCheckChildIndependently(fn)`, so if we have a `UnresolvedDotExpr` with a `subscript` base, we can be sure the problem is that there is no such member. If there is a member, type-checking will not fail (fnExpr != nullptr) and the expression will be further diagnosed without entering the fixit logic (e.g `cannot call value of non-function type`). `visitApplyExpr` secures we are talking about a function call.

Reply:
I think we can emit an error with or without a note from `diagnoseSubscriptMisuse`, because conditions are checked there anyway, and it's totally fine it diagnose this problem in different places because not all of the expressions which have this problem going to be applications I think (I wonder what happens if we just pass it to a function or use it as `.subscript` contextually somewhere).

I'm not a big fan of such a loose coupling between a diagnostic and its note, because though `typeCheckChildIndependently` would (most likely) produce a diagnostic there is no good way to say what did it produce, and if `diagnoseUnviableLookupResults` changes in any way in the future we might end up attaching note to something completely unrelated.

Reply:
Let's diagnose the error completely in `diagnoseSubscriptMisuse`, so it's emitted anyways if it has to be emitted, without falling through into type-check. It's a matter of verifying there aren't any viable member look up results. I will remove what I added to `diagnoseUnviableLookupResults`, unite the error message with the note and have a new note that will instead point where the resolved subscript is declared if we add the fixits. `; did you mean to use the subscript operator?`,  `declared here`.

Reply:
Sure, that sounds good to me! I just wanted to point out that member reference is not necessary only happens in context of function application so it might be useful to have it in `diagnoseUnviableLookupResults` still...

Reply:
I'll leave it just in case then.